src/passes/pass-utils.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_passes_pass_utils_h
+#define wasm_passes_pass_utils_h
+
+#include <unordered_set>
+
+#include <pass.h>
+#include <wasm.h>
+
+namespace wasm {
+
+namespace PassUtils {
+
+using FuncSet = std::unordered_set<Function*>;
+
+// A wrapper around a parallel pass that filters it to run run only on select
+// functions.
+struct FilteredPass : public Pass {
+  std::unique_ptr<Pass> create() override {
+    // Function-parallel passes get a new instance per function. Create a copy
+    // of the wrapped pass along with ourselves.
+    return std::make_unique<FilteredPass>(pass->create(), relevantFuncs);
+  }
+
+  FilteredPass(std::unique_ptr<Pass> pass, const FuncSet& relevantFuncs)
+    : pass(std::move(pass)), relevantFuncs(relevantFuncs) {}
+
+  bool isFunctionParallel() override {
+    assert(pass->isFunctionParallel());
+    return true;
+  }
+
+  void runOnFunction(Module* module, Function* func) override {
+    if (!relevantFuncs.count(func)) {
```

Comment:
To verify this PR is correct (on top of the test suite having no changes), that is, that it actually runs only on the filtered functions, I checked what happens when this condition is flipped so that we operate on the inverse set of the functions we should. As expected the test suite then gets some huge diffs in asyncify and inlining tests.